state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -250,40 +249,62 @@ impl EventSubscriptionService {
 
     /// This notifies all the reconfiguration subscribers of the on-chain
     /// configurations at the specified version.
-    fn notify_reconfiguration_subscribers(&mut self, version: Version) -> Result<(), Error> {
+    fn notify_reconfiguration_subscribers(
+        &mut self,
+        config_registry: &[ConfigID],
+        version: Version,
+    ) -> Result<(), Error> {
         if self.reconfig_subscriptions.is_empty() {
             return Ok(()); // No reconfiguration subscribers!
         }
 
-        let new_configs = self.read_on_chain_configs(version)?;
+        let new_configs = self.read_on_chain_configs(config_registry, version)?;
         for (_, reconfig_subscription) in self.reconfig_subscriptions.iter_mut() {
             reconfig_subscription.notify_subscriber_of_configs(version, new_configs.clone())?;
         }
 
         Ok(())
     }
 
-    /// Reads the values of all on-chain configs at the specified version.
-    fn read_on_chain_configs(&self, version: Version) -> Result<OnChainConfigPayload, Error> {
-        // Get the access paths for all the configs in the registry
-        let config_access_paths = ON_CHAIN_CONFIG_REGISTRY
-            .iter()
-            .map(|config_id| config_id.access_path())
-            .collect();
-
-        // Read the config values
-        let on_chain_configs = self
-            .storage
-            .read()
-            .reader
-            .deref()
-            .batch_fetch_resources_by_version(config_access_paths, version)
-            .map_err(|error| {
-                Error::UnexpectedErrorEncountered(format!(
-                    "Failed to batch fetch resources at version: {:?}. Error: {:?}",
-                    version, error
-                ))
-            })?;
+    /// Fetches the given config IDs on-chain at the specified version.
+    /// Note: We cannot assume that all configs will exist on-chain. As such, we
+    /// must fetch each resource one at a time. Reconfig subscribers must be able
+    /// to handle on-chain configs not existing in a reconfiguration notification.
+    fn read_on_chain_configs(
+        &self,
+        config_registry: &[ConfigID],
+        version: Version,
+    ) -> Result<OnChainConfigPayload, Error> {
+        // Build a map from config ID to the config value found on-chain
+        let mut config_id_to_config = HashMap::new();
+        for config_id in config_registry.iter() {
+            if let Ok(config_list) = self
+                .storage
+                .read()
+                .reader
+                .deref()
+                .batch_fetch_resources_by_version(vec![config_id.access_path()], version)
```

Comment:
Should we change this interface? Rather than failing if anything is missing, it could return a `Result<Vec<Option<Vec<u8>>>>`. It would return `Err` if there was a DB error or deserialization error. An entry would be `None` if there was no account or no resource.

Reply:
This is something I'm planning on doing in the future. For now, though, I've already created a bootcamp task as a way to help onboard someone. We can then take it from there regarding the best way to fix this up 😄